<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.36">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>SingleCell RNA-seq Workshop - sc_integration</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">SingleCell RNA-seq Workshop</span>
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./sc_scripts.html">General Scripts</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./sc_integrating.html" aria-current="page">Integration Scripts</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./sc_metacell.html">Metacell Scripts</a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#data-integration" id="toc-data-integration" class="nav-link active" data-scroll-target="#data-integration">Data Integration</a>
  <ul class="collapse">
  <li><a href="#evaluating-integration" id="toc-evaluating-integration" class="nav-link" data-scroll-target="#evaluating-integration">Evaluating integration</a></li>
  </ul></li>
  <li><a href="#differential-expression" id="toc-differential-expression" class="nav-link" data-scroll-target="#differential-expression">Differential Expression</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">sc_integration</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"pacman"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"pacman"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"SingleCellExperiment"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"Seurat"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"hdf5r"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"DropletUtils"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"scater"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"Matrix"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"scran"</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"igraph"</span>)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"SingleR"</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"celldex"</span>)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"iSEE"</span>)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"Nebulosa"</span>)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"UpSetR"</span>)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"clustree"</span>)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"Rmagic"</span>)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">"SeuratWrappers"</span>)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co">#p_load("harmony")</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"devtools"</span>))</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"devtools"</span>)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">"findPC"</span>))</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"haotian-zhuang/findPC"</span>)</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">"SeuratWrappers"</span>))</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">'satijalab/seurat-wrappers'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="data-integration" class="level2">
<h2 class="anchored" data-anchor-id="data-integration">Data Integration</h2>
<section id="pre-process" class="level4">
<h4 class="anchored" data-anchor-id="pre-process">Pre-process</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"pbmc_1k_v2_filtered_feature_bc_matrix.tar.gz"</span>)){</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">"wget -q https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v2/pbmc_1k_v2_filtered_feature_bc_matrix.tar.gz &amp;&amp; tar xzvf pbmc_1k_v2_filtered_feature_bc_matrix.tar.gz &amp;&amp; mkdir -p chem2 &amp;&amp; mv filtered_feature_bc_matrix chem2"</span>)}</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"pbmc_1k_v3_filtered_feature_bc_matrix.tar.gz"</span>)){</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">"wget -q https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_filtered_feature_bc_matrix.tar.gz &amp;&amp; tar xzvf pbmc_1k_v3_filtered_feature_bc_matrix.tar.gz &amp;&amp; mkdir -p chem3 &amp;&amp; mv filtered_feature_bc_matrix chem3"</span>)}</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>mat<span class="ot">&lt;-</span>Seurat<span class="sc">::</span><span class="fu">Read10X</span>(<span class="at">data.dir =</span> <span class="st">"chem2/filtered_feature_bc_matrix/"</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>chem2<span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> mat, <span class="at">project =</span> <span class="st">"chem2"</span>, <span class="at">min.cells =</span> <span class="dv">0</span>, <span class="at">min.features =</span><span class="dv">0</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>mat<span class="ot">&lt;-</span>Seurat<span class="sc">::</span><span class="fu">Read10X</span>(<span class="at">data.dir =</span>  <span class="st">"chem3/filtered_feature_bc_matrix/"</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>chem3<span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> mat, <span class="at">project =</span> <span class="st">"chem3"</span>, <span class="at">min.cells =</span> <span class="dv">0</span>, <span class="at">min.features =</span><span class="dv">0</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>chem2[[<span class="st">"Mitochondrial"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(chem2, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>qc.lib2 <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">isOutlier</span>(chem2<span class="sc">$</span>nCount_RNA, <span class="at">log=</span><span class="cn">TRUE</span>, <span class="at">type=</span><span class="st">"lower"</span>,<span class="at">nmads=</span><span class="dv">3</span>) <span class="co">#sce_data$sum</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>qc.nexprs2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(chem2<span class="sc">$</span>nFeature_RNA, <span class="at">log=</span><span class="cn">TRUE</span>,<span class="at">nmads=</span><span class="dv">3</span>, <span class="at">type=</span><span class="st">"lower"</span>) <span class="co">#sce_data$detected</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>qc.mito2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(chem2<span class="sc">$</span>Mitochondrial ,<span class="at">nmads=</span><span class="dv">3</span>,<span class="at">type=</span><span class="st">"higher"</span>) <span class="co">#sce_data$subsets_Mitochondrial_percent</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>discard2 <span class="ot">&lt;-</span> qc.lib2 <span class="sc">|</span> qc.nexprs2  <span class="sc">|</span> qc.mito2</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>chem2<span class="sc">$</span>discard<span class="ot">&lt;-</span>discard2</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>chem2<span class="ot">&lt;-</span><span class="fu">subset</span>(chem2,<span class="at">subset=</span> discard<span class="sc">!=</span><span class="cn">TRUE</span>)</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>chem3[[<span class="st">"Mitochondrial"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(chem3, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>qc.lib2 <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">isOutlier</span>(chem3<span class="sc">$</span>nCount_RNA, <span class="at">log=</span><span class="cn">TRUE</span>, <span class="at">type=</span><span class="st">"lower"</span>,<span class="at">nmads=</span><span class="dv">3</span>) <span class="co">#sce_data$sum</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>qc.nexprs2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(chem3<span class="sc">$</span>nFeature_RNA, <span class="at">log=</span><span class="cn">TRUE</span>,<span class="at">nmads=</span><span class="dv">3</span>, <span class="at">type=</span><span class="st">"lower"</span>) <span class="co">#sce_data$detected</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>qc.mito2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(chem3<span class="sc">$</span>Mitochondrial ,<span class="at">nmads=</span><span class="dv">3</span>,<span class="at">type=</span><span class="st">"higher"</span>) <span class="co">#sce_data$subsets_Mitochondrial_percent</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>discard2 <span class="ot">&lt;-</span> qc.lib2 <span class="sc">|</span> qc.nexprs2  <span class="sc">|</span> qc.mito2</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>chem3<span class="sc">$</span>discard<span class="ot">&lt;-</span>discard2</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>chem3<span class="ot">&lt;-</span><span class="fu">subset</span>(chem3,<span class="at">subset=</span> discard<span class="sc">!=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="join-cells" class="level4">
<h4 class="anchored" data-anchor-id="join-cells">Join Cells</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>seurat_data_merged<span class="ot">&lt;-</span><span class="fu">merge</span>(chem2,chem3,<span class="at">add.cell.ids =</span> <span class="fu">c</span>(<span class="st">"V2"</span>, <span class="st">"V3"</span>), <span class="at">project =</span> <span class="st">"merged"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>seurat_data_merged <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(seurat_data_merged, <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, <span class="at">scale.factor =</span> <span class="dv">10000</span>, <span class="at">verbose=</span>F)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>seurat_data_merged <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(seurat_data_merged, <span class="at">assay=</span><span class="st">"RNA"</span>,<span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>, <span class="at">verbose=</span>F)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>seurat_data_merged <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seurat_data_merged, <span class="at">features =</span> <span class="fu">rownames</span>(seurat_data_merged))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>seurat_data_merged <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seurat_data_merged,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> seurat_data_merged),</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">verbose =</span> F,<span class="at">npcs =</span> <span class="dv">50</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>seurat_data_merged <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seurat_data_merged,<span class="at">reduction =</span> <span class="st">"pca"</span>,<span class="at">verbose =</span> F,<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">11</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_data_merged, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">group.by =</span> <span class="st">"orig.ident"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="integration-seurat" class="level4">
<h4 class="anchored" data-anchor-id="integration-seurat">Integration Seurat</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>seurat.list<span class="ot">&lt;-</span><span class="fu">list</span>(chem2,chem3)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>seurat.list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> seurat.list, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(x,<span class="at">verbose=</span>F)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(x, <span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>,<span class="at">verbose=</span>F)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># select features that are repeatedly variable across datasets for integration</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>features <span class="ot">&lt;-</span> <span class="fu">SelectIntegrationFeatures</span>(<span class="at">object.list =</span> seurat.list)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(features)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GNLY"   "IGLC2"  "IGKC"   "GZMB"   "S100A8" "S100A9"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(features)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2000</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>anchors <span class="ot">&lt;-</span> <span class="fu">FindIntegrationAnchors</span>(<span class="at">object.list =</span> seurat.list, <span class="at">anchor.features =</span> features,<span class="at">verbose =</span> F)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">IntegrateData</span>(<span class="at">anchorset =</span> anchors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging dataset 1 into 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting anchors for merged samples</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Integrating data</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seurat_combined, <span class="at">features =</span> <span class="fu">rownames</span>(seurat_combined))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seurat_combined,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> seurat_combined),</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">verbose =</span> F,<span class="at">npcs =</span> <span class="dv">50</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seurat_combined,<span class="at">reduction =</span> <span class="st">"pca"</span>,<span class="at">verbose =</span> F,<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">11</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(seurat_combined, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seurat_combined, <span class="at">resolution =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 2005
Number of edges: 92431

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9063
Number of communities: 11
Elapsed time: 0 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_combined, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">group.by =</span> <span class="st">"orig.ident"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_combined, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">split.by =</span> <span class="st">"orig.ident"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>seurat_combined</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
35538 features across 2005 samples within 2 assays 
Active assay: integrated (2000 features, 2000 variable features)
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Integrated values should not be used for marker identification. It is important, in case of Seurat Object, to set the default assay again as the normalized values.</p>
</div>
</div>
</section>
<section id="integration-harmony" class="level4">
<h4 class="anchored" data-anchor-id="integration-harmony">Integration Harmony</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>seurat_data_harmony<span class="ot">&lt;-</span>harmony<span class="sc">::</span><span class="fu">RunHarmony</span>(seurat_data_merged,<span class="at">group.by.vars =</span> <span class="st">"orig.ident"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 1/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 2/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 3/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 4/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony converged after 4 iterations</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Invalid name supplied, making object name syntactically valid. New
object name is Seurat..ProjectDim.RNA.harmony; see ?make.names for more details
on syntax validity</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>seurat_data_harmony <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seurat_data_harmony,<span class="at">reduction =</span> <span class="st">"harmony"</span>,<span class="at">verbose =</span> F,<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">11</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_data_harmony, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">group.by =</span> <span class="st">"orig.ident"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="evaluating-integration" class="level3">
<h3 class="anchored" data-anchor-id="evaluating-integration">Evaluating integration</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>seurat.list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> seurat.list, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    x<span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(x, <span class="at">features =</span> <span class="fu">rownames</span>(x))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(x,<span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> x),</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">verbose =</span> F,<span class="at">npcs =</span> <span class="dv">50</span>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(x,<span class="at">reduction =</span> <span class="st">"pca"</span>,<span class="at">verbose =</span> F,<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">11</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(x, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(x, <span class="at">resolution =</span> <span class="fl">0.5</span>)</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(x, <span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>,<span class="at">verbose=</span>F)</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 927
Number of edges: 45294

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8671
Number of communities: 7
Elapsed time: 0 seconds</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 1078
Number of edges: 41871

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8839
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>tab<span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">before=</span>seurat.list[[<span class="dv">1</span>]]<span class="sc">$</span>seurat_clusters, <span class="at">after=</span>seurat_combined[,<span class="fu">colnames</span>(seurat.list[[<span class="dv">1</span>]])][[]]<span class="sc">$</span>seurat_clusters) <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(<span class="fu">log2</span>(tab<span class="sc">+</span><span class="dv">1</span>), <span class="at">color=</span><span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))(<span class="dv">101</span>),<span class="at">cluster_rows =</span> F,<span class="at">cluster_cols =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>tab<span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">before=</span>seurat.list[[<span class="dv">2</span>]]<span class="sc">$</span>seurat_clusters, <span class="at">after=</span>seurat_combined[,<span class="fu">colnames</span>(seurat.list[[<span class="dv">2</span>]])][[]]<span class="sc">$</span>seurat_clusters) <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(<span class="fu">log2</span>(tab<span class="sc">+</span><span class="dv">1</span>), <span class="at">color=</span><span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))(<span class="dv">101</span>),<span class="at">cluster_rows =</span> F,<span class="at">cluster_cols =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-10-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_combined) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_combined) <span class="ot">&lt;-</span> seurat_combined<span class="sc">$</span>seurat_clusters</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>conserved.markers <span class="ot">&lt;-</span> <span class="fu">FindConservedMarkers</span>(seurat_combined, <span class="at">ident.1 =</span> <span class="dv">0</span>, <span class="at">grouping.var =</span> <span class="st">"orig.ident"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(conserved.markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          chem2_p_val chem2_avg_log2FC chem2_pct.1 chem2_pct.2 chem2_p_val_adj
S100A12 3.003171e-177         5.122899       0.988       0.044   1.007203e-172
VCAN    1.115710e-165         3.402782       0.912       0.021   3.741867e-161
CD14    1.741667e-113         2.201955       0.680       0.018   5.841202e-109
MNDA    2.973783e-152         3.370733       0.932       0.065   9.973475e-148
MS4A6A  1.950469e-149         2.957007       0.876       0.034   6.541484e-145
S100A8  6.811264e-146         6.405144       1.000       0.216   2.284362e-141
          chem3_p_val chem3_avg_log2FC chem3_pct.1 chem3_pct.2 chem3_p_val_adj
S100A12 2.358008e-213         4.469216       0.987       0.025   7.908287e-209
VCAN    7.530415e-206         3.974759       0.990       0.048   2.525551e-201
CD14    2.686868e-202         2.871006       0.967       0.038   9.011218e-198
MNDA    4.119775e-194         3.409500       1.000       0.083   1.381690e-189
MS4A6A  4.972831e-194         2.830633       0.974       0.045   1.667788e-189
S100A8  7.148624e-194         6.448778       1.000       0.108   2.397506e-189
             max_pval minimump_p_val
S100A12 3.003171e-177  4.716016e-213
VCAN    1.115710e-165  1.506083e-205
CD14    1.741667e-113  5.373736e-202
MNDA    2.973783e-152  8.239549e-194
MS4A6A  1.950469e-149  9.945662e-194
S100A8  6.811264e-146  1.429725e-193</code></pre>
</div>
</div>
</section>
</section>
<section id="differential-expression" class="level2">
<h2 class="anchored" data-anchor-id="differential-expression">Differential Expression</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>seurat_combined<span class="sc">$</span>clust_orig.ident<span class="ot">&lt;-</span> <span class="fu">paste</span>(seurat_combined<span class="sc">$</span>seurat_clusters,seurat_combined<span class="sc">$</span>orig.ident, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_combined) <span class="ot">&lt;-</span> <span class="st">"clust_orig.ident"</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>dfex_clust0<span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_combined, <span class="at">ident.1 =</span> <span class="st">"0_chem2"</span>, <span class="at">ident.2 =</span> <span class="st">"0_chem3"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>,<span class="at">only.pos =</span> F)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dfex_clust0, <span class="at">n =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               p_val avg_log2FC pct.1 pct.2    p_val_adj
RPL21   3.961619e-91  1.5783971 1.000 1.000 1.328648e-86
RPL7    1.727843e-90  1.9019795 1.000 0.997 5.794838e-86
MT-ND1  8.752324e-90 -1.7221381 0.988 0.997 2.935355e-85
RPL13A  1.078006e-89  2.1108584 1.000 0.987 3.615415e-85
MT-ATP6 1.463502e-87 -1.5094426 0.996 0.997 4.908293e-83
RPS27   9.397522e-86  1.2758131 1.000 1.000 3.151741e-81
RPL27A  7.646383e-85  2.0900508 0.992 0.967 2.564444e-80
TMSB4X  1.067049e-83  1.0082501 1.000 1.000 3.578668e-79
RPL34   3.778420e-82  1.2388800 1.000 1.000 1.267207e-77
RPS17   6.336518e-82  2.8078563 0.968 0.683 2.125142e-77
RPLP2   1.131240e-81  1.4007701 1.000 0.993 3.793954e-77
RPS15   1.567411e-80  1.2551856 1.000 1.000 5.256783e-76
MT-CO3  6.279638e-80 -0.9356098 0.996 0.997 2.106065e-75
RPL39   9.870788e-79  0.8841529 1.000 1.000 3.310465e-74
MT-CO2  2.136420e-76 -0.8839316 1.000 0.997 7.165126e-72</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat_combined, <span class="at">features =</span> <span class="st">"CRIP1"</span>, <span class="at">split.plot =</span> T,<span class="at">split.by =</span> <span class="st">"orig.ident"</span>, <span class="at">group.by =</span> <span class="st">"seurat_clusters"</span>,<span class="at">pt.size =</span> <span class="fl">0.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The default behaviour of split.by has changed.
Separate violin plots are now plotted side-by-side.
To restore the old behaviour of a single split violin,
set split.plot = TRUE.
      
This message will be shown once per session.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>seurat_combined<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>data[<span class="st">"SUB1"</span>,] <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"Cell"</span>) <span class="sc">%&gt;%</span> <span class="fu">left_join</span>(<span class="fu">as_tibble</span>(seurat_combined[[]], <span class="at">rownames=</span><span class="st">"Cell"</span>), <span class="at">by=</span><span class="fu">c</span>(<span class="st">"Cell"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(seurat_clusters<span class="sc">==</span><span class="dv">0</span>)<span class="sc">%&gt;%</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>seurat_clusters, <span class="at">y=</span>value))<span class="sc">+</span><span class="fu">geom_violin</span>(<span class="fu">aes</span>(<span class="at">fill=</span>orig.ident),<span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">adjust=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>    <span class="co">#ggbeeswarm::geom_quasirandom(size=0.1)+</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#theme(legend.position = "none")</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat_combined, <span class="at">features =</span> <span class="st">"ISG15"</span>, <span class="at">group.by =</span> <span class="st">"seurat_clusters"</span>,<span class="at">pt.size =</span> <span class="fl">0.2</span>,<span class="at">fill.by =</span> <span class="st">"orig.ident"</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sc_integrating_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"Oshlack/speckle"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Skipping install of 'speckle' from a github remote, the SHA1 (83df3e9e) has not changed since last install.
  Use `force = TRUE` to force installation</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>